### PR TITLE
[WIP]Cross-Chain Stake

### PR DESCRIPTION
**Pull request by pythonberg1997**

### CODE REVIEW

File: contracts/CrossStake.sol [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    require(amount > oracleRelayerFee, "Send value cannot cover the relayer fee");
+
+    bytes memory msgBytes;
+    msgBytes = abi.encode(EVENT_STAKE, msg.sender, amount-oracleRelayerFee, validator);
+    ICrossChain.sendSynPackage(CROSS_STAKE_CHANNELID, msgBytes, oracleRelayerFee);
+  }
+
+  function unstake(address validator, uint256 amount) payable initRelayerFee {
+    uint256 _value = msg.value;
+    require(amount >= oracleRelayerFee, "Send value cannot cover the relayer fee");
```
Comment (gothery001) [inline contracts/CrossStake.sol]: `amount`  => _value ?

File: contracts/BSCValidatorSet.sol [hunk 2]
```diff
@@ -697,7 +697,7 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
     } else if (Memory.compareStrings(key, "finalityRewardRatio")) {
       require(value.length == 32, "length of finalityRewardRatio mismatch");
       uint256 newFinalityRewardRatio = BytesToTypes.bytesToUint256(32, value);
-      require(newFinalityRewardRatio >= 1 && newFinalityRewardRatio < 100, "the finality reward ratio out of range!");
+      require(newFinalityRewardRatio >= 1 && newFinalityRewardRatio < 100, "the finalityRewardRatio is out of range");
```
Comment (realuncle) [inline contracts/BSCValidatorSet.sol]: Please do not change unrelated code.

File: contracts/BSCValidatorSet.template [hunk 3]
```diff
@@ -698,7 +698,7 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
     } else if (Memory.compareStrings(key, "finalityRewardRatio")) {
       require(value.length == 32, "length of finalityRewardRatio mismatch");
       uint256 newFinalityRewardRatio = BytesToTypes.bytesToUint256(32, value);
-      require(newFinalityRewardRatio >= 1 && newFinalityRewardRatio < 100, "the finality reward ratio out of range!");
+      require(newFinalityRewardRatio >= 1 && newFinalityRewardRatio < 100, "the finalityRewardRatio is out of range");
```
Comment (realuncle) [inline contracts/BSCValidatorSet.template]: so do here.

File: contracts/CrossStake.sol [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+import "./interface/IParamSubscriber.sol";
+import "./interface/IApplication.sol";
+import "./lib/SafeMath.sol";
+import "./lib/RLPDecode.sol";
+
+contract CrossStake is System, IParamSubscriber, IApplication {
+  using SafeMath for uint256;
+  using RLPDecode for *;
+
+  uint8 constant public CROSS_STAKE_CHANNELID = 0x10;
```
Comment (realuncle) [inline contracts/CrossStake.sol]: This should in `system.sol`

File: contracts/CrossStake.sol [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+
+  // Cross-Chain Stake Event type
+  uint8 public constant EVENT_STAKE = 0x00;
```
Comment (realuncle) [inline contracts/CrossStake.sol]: Please do not start from `0x00`

File: contracts/CrossStake.sol [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+  uint8 public constant EVENT_UNSTAKE = 0x01;
+  uint8 public constant EVENT_CLAIM_REWARD = 0x02;
+  uint8 public constant EVENT_CLAIM_UNSTAKE = 0x03;
+  uint8 public constant EVENT_RESTAKE = 0x04;
+
+  //TODO
+  uint256 public constant INIT_ORACLE_RELAYER_FEE = 0;
```
Comment (realuncle) [inline contracts/CrossStake.sol]: The  INIT_ORACLE_RELAYER_FEE is `2000000000000000`.

File: contracts/CrossStake.sol [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  uint256 public constant INIT_BSC_RELAYER_FEE = 0;
```
Comment (realuncle) [inline contracts/CrossStake.sol:26]: Please make it same as token transfer from BC to BSC.

File: contracts/CrossStake.sol [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+
+  uint256 public oracleRelayerFee;
+  uint256 public BSCRelayerFee;
+
+  modifier initRelayerFee() {
+    if (oracleRelayerFee == 0) {
```
Comment (realuncle) [inline contracts/CrossStake.sol]: `oracleRelayerFee == 0 || BSCRelayerFee == 0`

File: contracts/CrossStake.sol [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+      oracleRelayerFee = INIT_ORACLE_RELAYER_FEE;
+      BSCRelayerFee = INIT_BSC_RELAYER_FEE;
+    }
+    _;
+  }
+
+  function stake(address validator) payable initRelayerFee {
+    uint256 amount = msg.value;
+    require(amount > oracleRelayerFee, "Send value cannot cover the relayer fee");
```
Comment (realuncle) [inline contracts/CrossStake.sol:40]: should be `>=`

File: contracts/CrossStake.sol [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+
+    bytes memory msgBytes;
+    msgBytes = abi.encode(EVENT_STAKE, msg.sender, amount-oracleRelayerFee, validator);
```
Comment (realuncle) [inline contracts/CrossStake.sol]: Please do not use abi encode but still follow rlp encode just like other cross chain app.

Comment (realuncle) [reply contracts/CrossStake.sol]: I will suggest the package as:
```
{
  EventType: uint8,
  EventInfo:    bytes, // rlp encoded.
}
```

File: contracts/CrossStake.sol [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+    ICrossChain.sendSynPackage(CROSS_STAKE_CHANNELID, msgBytes, oracleRelayerFee);
+  }
+
+  function unstake(address validator, uint256 amount) payable initRelayerFee {
+    uint256 _value = msg.value;
+    require(amount >= oracleRelayerFee, "Send value cannot cover the relayer fee");
+
+    bytes memory msgBytes;
+    msgBytes = abi.encode(EVENT_UNSTAKE, msg.sender, validator, amount);
+    ICrossChain.sendSynPackage(CROSS_STAKE_CHANNELID, msgBytes, oracleRelayerFee);
```
Comment (realuncle) [inline contracts/CrossStake.sol]: should be `    ICrossChain.sendSynPackage(CROSS_STAKE_CHANNELID, msgBytes, amount);`? 
Since all msg.value is used as fee.

File: contracts/CrossStake.sol [hunk 12]
```diff
@@ -0,0 +1,99 @@
...
+  }
+
+  function claimReward() payable initRelayerFee {
+    uint256 _value = msg.value;
+    require(amount >= oracleRelayerFee+BSCRelayerFee, "Send value cannot cover the relayer fee");
+
+    bytes memory msgBytes;
+    msgBytes = abi.encode(EVENT_CLAIM_REWARD, msg.sender, BSCRelayerFee);
```
Comment (realuncle) [inline contracts/CrossStake.sol]: `    msgBytes = abi.encode(EVENT_CLAIM_REWARD, msg.sender, amount-oracleRelayerFee);`?

File: contracts/CrossStake.sol [hunk 13]
```diff
@@ -0,0 +1,99 @@
...
+  function unstake(address validator, uint256 amount) payable initRelayerFee {
+    uint256 _value = msg.value;
+    require(amount >= oracleRelayerFee, "Send value cannot cover the relayer fee");
+
+    bytes memory msgBytes;
+    msgBytes = abi.encode(EVENT_UNSTAKE, msg.sender, validator, amount);
+    ICrossChain.sendSynPackage(CROSS_STAKE_CHANNELID, msgBytes, oracleRelayerFee);
+  }
+
+  function claimReward() payable initRelayerFee {
```
Comment (realuncle) [inline contracts/CrossStake.sol]: To avoid that the msg.sender is not payable, I will suggest `claimReward(address receiver)`, user can use any address to receive reward and unstake fund.

File: contracts/CrossStake.sol [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+    uint256 _value = msg.value;
+    require(amount >= oracleRelayerFee+BSCRelayerFee, "Send value cannot cover the relayer fee");
+
+    bytes memory msgBytes;
+    msgBytes = abi.encode(EVENT_CLAIM_REWARD, msg.sender, BSCRelayerFee);
+    ICrossChain.sendSynPackage(CROSS_STAKE_CHANNELID, msgBytes, oracleRelayerFee);
+  }
+
+  function claimUnstake() payable initRelayerFee {
```
Comment (realuncle) [inline contracts/CrossStake.sol]: `claimUnstake(address receiver)`

File: contracts/CrossStake.sol [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+    uint256 _value = msg.value;
+    require(amount >= oracleRelayerFee+BSCRelayerFee, "Send value cannot cover the relayer fee");
+
+    bytes memory msgBytes;
+    msgBytes = abi.encode(EVENT_CLAIM_UNSTAKE, msg.sender, BSCRelayerFee);
```
Comment (realuncle) [inline contracts/CrossStake.sol]: `    msgBytes = abi.encode(EVENT_CLAIM_UNSTAKE, msg.sender, amount-oracleRelayerFee);`

File: contracts/CrossStake.sol [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+    require(amount >= oracleRelayerFee, "Send value cannot cover the relayer fee");
+
+    bytes memory msgBytes;
+    msgBytes = abi.encode(EVENT_UNSTAKE, msg.sender, validator, amount);
+    ICrossChain.sendSynPackage(CROSS_STAKE_CHANNELID, msgBytes, oracleRelayerFee);
+  }
+
+  function claimReward() payable initRelayerFee {
+    uint256 _value = msg.value;
+    require(amount >= oracleRelayerFee+BSCRelayerFee, "Send value cannot cover the relayer fee");
```
Comment (realuncle) [inline contracts/CrossStake.sol]: typo here? amount should be msg.value.

**Comment by realuncle** [comment]
Missing testcase, we need some testcase to cover it.

File: contracts/CrossStake.sol (hunk 11 above)
Comment (gothery001) [reply contracts/CrossStake.sol]: > should be ` ICrossChain.sendSynPackage(CROSS_STAKE_CHANNELID, msgBytes, amount);`? Since all msg.value is used as fee.

should be ` ICrossChain.sendSynPackage(CROSS_STAKE_CHANNELID, msgBytes, _value);`

File: contracts/CrossStake.sol [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+  function stake(address validator) payable initRelayerFee {
+    uint256 amount = msg.value;
+    require(amount > oracleRelayerFee, "Send value cannot cover the relayer fee");
+
+    bytes memory msgBytes;
+    msgBytes = abi.encode(EVENT_STAKE, msg.sender, amount-oracleRelayerFee, validator);
+    ICrossChain.sendSynPackage(CROSS_STAKE_CHANNELID, msgBytes, oracleRelayerFee);
+  }
+
+  function unstake(address validator, uint256 amount) payable initRelayerFee {
```
Comment (gothery001) [inline contracts/CrossStake.sol]: should add `nonReentrant` even if there is no risk

File: contracts/CrossStake.sol (hunk 13 above)
Comment (gothery001) [inline contracts/CrossStake.sol]: should add `nonReentrant` even if there is no risk

File: contracts/CrossStake.sol (hunk 14 above)
Comment (gothery001) [inline contracts/CrossStake.sol]: should add `nonReentrant` even if there is no risk

File: contracts/CrossStake.sol (hunk 1 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: Fixed

File: contracts/BSCValidatorSet.sol (hunk 2 above)
Comment (pythonberg1997) [reply contracts/BSCValidatorSet.sol]: Done

File: contracts/BSCValidatorSet.template (hunk 3 above)
Comment (pythonberg1997) [reply contracts/BSCValidatorSet.template]: Done

File: contracts/CrossStake.sol (hunk 4 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: Fixed

File: contracts/CrossStake.sol (hunk 5 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: Fixed

File: contracts/CrossStake.sol (hunk 6 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: Fixed

File: contracts/CrossStake.sol (hunk 7 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol:26]: As I know, it's 0 right?

File: contracts/CrossStake.sol (hunk 8 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: Fixed

File: contracts/CrossStake.sol (hunk 10 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: The encode method has been fixed. But I don't understand the meaning of 
```
{
  EventType: uint8,
  EventInfo:    bytes, // rlp encoded.
}
```

File: contracts/CrossStake.sol (hunk 11 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: Fixed

File: contracts/CrossStake.sol (hunk 12 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: Fixed

File: contracts/CrossStake.sol (hunk 13 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: Done

File: contracts/CrossStake.sol (hunk 14 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: Done

File: contracts/CrossStake.sol (hunk 15 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: Done

File: contracts/CrossStake.sol (hunk 16 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: Done

File: contracts/CrossStake.sol (hunk 17 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: Done

File: contracts/CrossStake.sol (hunk 13 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: Done

File: contracts/CrossStake.sol (hunk 14 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: Done

File: contracts/CrossStake.sol [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+    uint256 _amount = msg.value;
+    require(amount >= oracleRelayerFee, "Send value cannot cover the relayer fee");
+
+    bytes[] memory elements = new bytes[](4);
+    elements[0] = EVENT_RESTAKE.encodeUint();
+    elements[1] = msg.sender.encodeAddress();
+    elements[2] = validator.encodeAddress();
+    elements[3] = amount.encodeUint();
+    bytes memory msgBytes = elements.encodeList();
+    ICrossChain.sendSynPackage(CROSS_STAKE_CHANNELID, msgBytes, _amount);
```
Comment (owen-reorg) [inline contracts/CrossStake.sol]: Why use both `oracleRelayerFee` and `amount` instead of just `amount` for the 3rd parameter?

File: contracts/CrossStake.sol [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+  function unstake(address validator, uint256 amount) payable initRelayerFee {
+    uint256 _amount = msg.value;
+    require(_amount >= oracleRelayerFee, "Send value cannot cover the relayer fee");
+
+    bytes[] memory elements = new bytes[](4);
+    elements[0] = EVENT_UNSTAKE.encodeUint();
+    elements[1] = msg.sender.encodeAddress();
+    elements[2] = validator.encodeAddress();
+    elements[3] = amount.encodeUint();
+    bytes memory msgBytes = elements.encodeList();
```
Comment (owen-reorg) [inline contracts/CrossStake.sol:95]: why not use `msgBytes = abi.encode(EVENT_UNSTAKE, msg.sender, validator, amount);` like in the template?

File: contracts/BSCValidatorSet.sol [hunk 20]
```diff
@@ -33,7 +33,8 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
   uint256 public constant EXPIRE_TIME_SECOND_GAP = 1000;
   uint256 public constant MAX_NUM_OF_VALIDATORS = 41;
 
-  bytes public constant INIT_VALIDATORSET_BYTES = hex"f84580f842f840949fb29aac15b9a4b7f17c3385939b007540f4d791949fb29aac15b9a4b7f17c3385939b007540f4d791949fb29aac15b9a4b7f17c3385939b007540f4d79164";
+  bytes public constant INIT_VALIDATORSET_BYTES = hex"f87680f873f871949fb29aac15b9a4b7f17c3385939b007540f4d791949fb29aac15b9a4b7f17c3385939b007540f4d791949fb29aac15b9a4b7f17c3385939b007540f4d79164b085e6972fc98cd3c81d64d40e325acfed44365b97a7567a27939c14dbc7512ddcf54cb1284eb637cfa308ae4e00cb5588";
```
Comment (NashBC) [inline contracts/BSCValidatorSet.sol]: Do we need to change this?

File: contracts/CrossStake.sol [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+    uint256 amount = msg.value;
+    require(amount > oracleRelayerFee, "Send value cannot cover the relayer fee or stake value is zero");
+
+    bytes[] memory elements = new bytes[](4);
+    elements[0] = EVENT_STAKE.encodeUint();
+    elements[1] = msg.sender.encodeAddress();
+    elements[2] = validator.encodeAddress();
+    elements[3] = amount.encodeUint();
+    bytes memory msgBytes = elements.encodeList();
+    ICrossChain.sendSynPackage(CROSS_STAKE_CHANNELID, msgBytes, oracleRelayerFee);
```
Comment (NashBC) [inline contracts/CrossStake.sol]: emit a stake event here?

Comment (NashBC) [reply contracts/CrossStake.sol]: Where does the staked BNB go?

Comment (NashBC) [reply contracts/CrossStake.sol]: Is it ok if we don't track how much BNB the msg.sender has staked?

File: contracts/CrossStake.sol [hunk 22]
```diff
@@ -0,0 +1,126 @@
...
+  }
+
+  function unstake(address validator, uint256 amount) payable initRelayerFee {
+    uint256 _amount = msg.value;
```
Comment (NashBC) [inline contracts/CrossStake.sol]: How can the users find how much BNB they have staked? Then they can make sure the amount will be no larger than that value, otherwise, they have paid the relayer fee, but the unstake failed.

File: contracts/CrossStake.sol [hunk 23]
```diff
@@ -0,0 +1,126 @@
...
+    require(_amount >= oracleRelayerFee, "Send value cannot cover the relayer fee");
+
+    bytes[] memory elements = new bytes[](4);
+    elements[0] = EVENT_UNSTAKE.encodeUint();
+    elements[1] = msg.sender.encodeAddress();
+    elements[2] = validator.encodeAddress();
+    elements[3] = amount.encodeUint();
+    bytes memory msgBytes = elements.encodeList();
+    ICrossChain.sendSynPackage(CROSS_STAKE_CHANNELID, msgBytes, _amount);
```
Comment (NashBC) [inline contracts/CrossStake.sol]: emit event also

File: contracts/CrossStake.sol [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+    elements[1] = receiver.encodeAddress();
+    uint256 _bSCRelayerFee = amount-oracleRelayerFee;
+    elements[2] = _bSCRelayerFee.encodeUint();
+    bytes memory msgBytes = elements.encodeList();
+    ICrossChain.sendSynPackage(CROSS_STAKE_CHANNELID, msgBytes, oracleRelayerFee);
+  }
+
+  function reStake(address validator, uint256 amount) payable initRelayerFee {
+    uint256 _amount = msg.value;
+    require(amount >= oracleRelayerFee, "Send value cannot cover the relayer fee");
```
Comment (NashBC) [inline contracts/CrossStake.sol]: I think here is _amount not amount

File: contracts/CrossStake.sol (hunk 18 above)
Comment (NashBC) [reply contracts/CrossStake.sol]: _amount -> oracleRelayerFee

File: contracts/System.sol [hunk 25]
```diff
@@ -30,6 +30,7 @@ contract System {
   address public constant TOKEN_MANAGER_ADDR = 0x0000000000000000000000000000000000001008;
   address public constant CROSS_CHAIN_CONTRACT_ADDR = 0x0000000000000000000000000000000000002000;
 
+  uint8 constant public CROSS_STAKE_CHANNELID = 0x10;
```
Comment (NashBC) [inline contracts/System.sol:33]: Move to the upper where defining the CHANNELID is better?

File: contracts/CrossStake.sol (hunk 18 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: It's a typo here, should be `require(_amount >= oracleRelayerFee, "Send value cannot cover the relayer fee")`.

Comment (pythonberg1997) [reply contracts/CrossStake.sol]: Sometimes all the `msg.value` will served as `oracleRelayerFee` and sometimes served as `oracleRelayerFee` and `BSCRelayerFee`. It depends on weather the action is one-pass communication or two-pass communication.

Comment (pythonberg1997) [reply contracts/CrossStake.sol]: In either case, all the `msg.value` should be sent even if it is greater than base fee.

File: contracts/CrossStake.sol (hunk 19 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol:95]: It's walt's advice. I forget to update the template.

File: contracts/BSCValidatorSet.sol (hunk 20 above)
Comment (pythonberg1997) [reply contracts/BSCValidatorSet.sol]: Fixed

File: contracts/CrossStake.sol (hunk 24 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: Fixed

File: contracts/System.sol (hunk 25 above)
Comment (pythonberg1997) [reply contracts/System.sol:33]: I'm afraid this will change the storage

File: contracts/CrossStake.sol (hunk 21 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: The staked BNB will go to BC side.

File: contracts/CrossStake.sol (hunk 22 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: As we discueed before, the balance should tracked by the project side. The users will know their staked amount by the LP tokens. If the amount is greater than the staked amount, maybe we should just unstake all the balance?

File: contracts/CrossStake.sol (hunk 23 above)
Comment (pythonberg1997) [reply contracts/CrossStake.sol]: Done
